xdsl/traits.py

```diff
@@ -94,6 +94,66 @@ def verify(self, op: Operation) -> None:
                 )
 
 
+class SingleBlockImplicitTerminator(OpTrait):
+    """
+    Checks and adds if missing the specified terminator to an operation which has
+    single-block region.
+
+    This should be fully compatible with MLIR's Trait:
+    https://mlir.llvm.org/docs/Traits/#single-block-with-implicit-terminator
+    """
+
+    parameters: type[Operation]
+
+    def verify(self, op: Operation) -> None:
+        for region in op.regions:
+            if len(region.blocks) > 1:
+                raise VerifyException(
+                    f"'{op.name}' does not contain single-block regions"
+                )
+            for block in region.blocks:
+                if (last_op := block.last_op) is None:
+                    raise VerifyException(
+                        f"'{op.name}' contains empty block instead of at least "
+                        f"terminating with {self.parameters.name}"
+                    )
+
+                if not isinstance(last_op, self.parameters):
+                    raise VerifyException(
+                        f"'{op.name}' terminates with operation {last_op.name} "
+                        f"instead of {self.parameters.name}"
+                    )
+
+
+def ensure_terminator(trait: SingleBlockImplicitTerminator, op: Operation) -> None:
```

Comment:
Maybe add a comment on what this exactly does ( I guess it first verifies that a terminator can be added and then does it, if the check succeeded?)